lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -511,6 +563,49 @@ class TraceProcessor {
     return {lcp, invalidated};
   }
 
+  /**
+   * @param {{frame: string, url: string, parent?: string}[]} frames
+   * @return {Map<string, string>}
+   */
+  static resolveRootFrames(frames) {
+    /** @type {Map<string, string>} */
+    const parentFrames = new Map();
+    for (const frameData of frames) {
+      if (!frameData.parent) continue;
+      parentFrames.set(frameData.frame, frameData.parent);
+    }
+
+    /** @type {Map<string, string>} */
+    const rootFrames = new Map();
+
+    /**
+     * @param {string} frame
+     * @return {string}
+     */
+    function resolveRootFrame(frame) {
+      let rootFrame = rootFrames.get(frame);
+      if (!rootFrame) {
+        const parentFrame = parentFrames.get(frame);
+        if (!parentFrame) {
+          rootFrames.set(frame, frame);
+          return frame;
+        }
+        rootFrame = rootFrames.get(parentFrame) || resolveRootFrame(parentFrame);
+        rootFrames.set(frame, rootFrame);
+      }
+      return rootFrame;
+    }
+
+    for (const frameData of frames) {
+      resolveRootFrame(frameData.frame);
+
+      // Early exit if map is filled in.
+      if (rootFrames.size === frames.length) break;
+    }
+
```

Comment:
I think finding the root frame could be simpler. I haven't tried this code, but this is what I'm thinking:

```
/** @type {Map<string, {id: string, url: string, parent?: FrameData}>} */
const frames = new Map();
// 1 ... initialize the map from the input param ...
// 2 ... loop the initial param again to set the parent of each value in the map ....

// 3 create a new map (frame id => root frame id)
/** @type {Map<string, string>} */
let frameIdToRootFrameId = new Map();
for (const frameData of frames.values()) {
  let cur = frameData;
  while (frameData.parent) {
    cur = cur.parent;
  }
  const rootFrameId = cur.id;
  frameIdToRootFrameId.set(frameData.id, rootFrameId);
}

return frameIdToRootFrameId;
```

The key is making an intermediate structure, and leverage regular object properties referencing other objects to model the hierarchy.

Reply:
The reason I used the recursive solution was so we wouldn't have to walk up the entire tree for every node. The code could short circuit if the root frame of the parent has already been found. 

Is that a performance benefit worth pursuing at the cost of code readability?

Reply:
We're talking about at most what, dozens of frames here? How much would cnn.com be? There'd be no performance issue at that scale.

I suppose the code I provided could also introduce a shortcut. For each iteration of the inner while loop, see if `cur.id` is a key in `frameIdToRootFrameId`. If so, that means the current `frameData` is a child of a previously processed frame, and so the roots are equivalent.

Reply:
CNN only has 20 so I think your code will be good